hummingbird/ml/operator_converters/sparkml/sql_transformer.py

```diff
@@ -0,0 +1,116 @@
+        self.transformer = True
+        self.input_names = input_names
+        self.project_list = project_list
+
+    def forward(self, *x):
+        inputs_dict = {name : tensor for name, tensor in zip(self.input_names, x)}
+        # The first node of the project_node_def is the name definition node. Hence ignored.
+        return [self.calculate_output_feature(project_node_def_stack[1:], inputs_dict)[0] for project_node_def_stack in self.project_list]
+
+    def calculate_output_feature(self, project_node_def_stack, inputs_dict):
+        if project_node_def_stack[0]['class'] == 'org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute':
+            return inputs_dict[project_node_def_stack[0]['nameParts'][1:-1].strip()], project_node_def_stack[1:]
+        elif project_node_def_stack[0]['class'] == 'org.apache.spark.sql.catalyst.expressions.Literal':
+            return float(project_node_def_stack[0]['value']), project_node_def_stack[1:]
+        elif project_node_def_stack[0]['class'] in self.supported_binary_ops:
+            op = self.supported_binary_ops[project_node_def_stack[0]['class']]
+            project_node_def_stack = project_node_def_stack[1:]
+            left, project_node_def_stack = self.calculate_output_feature(project_node_def_stack, inputs_dict)
+            right, project_node_def_stack = self.calculate_output_feature(project_node_def_stack, inputs_dict)
+            return op(left, right), project_node_def_stack
+        else:
+            raise RuntimeError('SQLTransformer encountered unsupported operator type: {}'.format(project_node_def_stack[0]['class']))
+
+
+def convert_sparkml_sql_transformer(operator, device, extra_config):
+    """
+    Converter for `pyspark.ml.feature.SQLTransformer`
+
+    Args:
+        operator: An operator wrapping a `pyspark.ml.feature.SQLTransformer`
+        device: String defining the type of device the converted operator should be run on
+        extra_config: Extra configuration used to select the best conversion strategy
+
+    Returns:
+        A PyTorch model
+    """
+    from pyspark.sql import SparkSession
+
+    spark = SparkSession.builder.getOrCreate()
+    parser = spark._jsparkSession.sessionState().sqlParser()
+    statement = operator.raw_operator.getStatement()
+    plan = parser.parsePlan(statement)
+    plan_json = json.loads(plan.toJSON())
+
+    project_node = [node for node in plan_json if node['class'] == 'org.apache.spark.sql.catalyst.plans.logical.Project']
+    if len(project_node) == 0:
+        raise RuntimeError('Couldn\'t find the projectList from the SQLTransformer statement: {}.'.format(statement))
+
+    # Ignores select *
+    project_list = [proj for proj in project_node[0]['projectList'] if proj[0]['class'] != 'org.apache.spark.sql.catalyst.analysis.UnresolvedStar']
+    input_names = [input.raw_name for input in operator.inputs]
+    return SQLTransformerModel(input_names, project_list, device)
```

Comment:
Ok so here you are generating a uber operator for the whole `SQLTransformer`. Another possibility is to use `SQLTransformer` as a specific case in `_parse` and add each single operator one by one into the topology. Have you thought about this other option? Do you have an idea on the possible pros and cons between these 2 possibilities?

Reply:
Right now the parsing logic is mostly focussed on topology of the DAG and the corresponding inputs and outputs. The semantics of the operator is handled at the actual operator definition. We can treat each project column as a different operator and handle it at the _parse level. But that will require elevating some of the operator level logic to the parsing logic. WDYT?

Reply:
I see your point. If we add the operator in _parse we are basically exploding it and merging the Spark SQL plan DAG into our topology DAG. This is similar to how we explode the `ColumnTransformer` or `Pipeline` object in sklearn models. So basically we need a parsing function specific for Spark SQL plans that (1) opens the plan, (2) traverse the plan, (3) transform each operator in the plan to a node in the topology object, and (4) wires the inputs\outpus.